### DISCUSSION TITLE
Quanto tempo você de fato codifica?

### DISCUSSION THREAD

**Discussion by danilovaz** [frontendbr/forum#136] [Discussão]
Essa era uma dúvida que eu tinha frequentemente. E não digo em termos de "banco de horas" mas quanto tempo, de fato, do meu dia eu passo programando? E em quais projetos? Quais as linguagens/tecnologias que eu mais utilizo? Porque eu passo muito tempo pesquisando e estudando também, queria uma margem de quanto tempo realmente fico com a mão na massa.

Jamais que eu iria fazer isso manualmente ou qualquer sistema que eu tenha que ficar colocando "Iniciei agora em projeto X" "Fui pra projeto Y" "Voltei pra projeto X"

Foi aí que encontrei o [Wakatime](https://wakatime.com) basicamente eu conecto ele com o meu Sublime e então ele identifica cada folder pai que eu abro como sendo um projeto e então começa a analisar projeto a projeto: quanto tempo fico em cada, porcentagem de arquivos que mais trabalho, média diária de trabalho, etc.

Um exemplo do meu

![imagem](http://i.picresize.com/images/2016/04/26/2xsDo.jpg)

Espero que possa ser útil a alguém também :smile:

**Comment by felipefialho** [comment]
Porra, bem interessante. 

Integrei no meu Sublime, vou testar :)

**Comment by RicardoJunior** [comment]
Opa! Eu uso o Wakatime pra saber os projetos e linguagens. Porém eu precisava de alguma coisa pra medir distrações, eu estava usando o RescueTime, na versão free.

https://www.rescuetime.com/

**Comment by woliveiras** [comment]
Eu coloquei no Sublimaum uma vez, mas, como mexo muito com WordPress, a galera ficou me zoando por aparecer um monte de PHP no relatório... Ai tirei pra parar o bullying. 😢

**Comment by felipefialho** [comment]
Atualiza em tempo real? Não tem nada no meu log do dia.

**Comment by danilovaz** [comment]
@LFeh Você instalou o package do seu sublime? 

Agora você tem que ir em `Preferences -> WakaTime -> Settings - User` e coloca a sua API Key (que você gera lá no site do Wakatime):

``` javascript
{
    // Your api key from https://wakatime.com/#apikey
    // Set this in your User specific WakaTime.sublime-settings file.
    "api_key": "APIKEY"
}
```

**Comment by danilovaz** [comment]
@woliveiras cyberbullyng com PHP kkkk. cara, tem como deixar o seu graph de linguagens e horas codando em modo privado, aí ninguém consegue ver, só você

**Comment by rubenmarcus** [comment]
wakatime é pago, ele funciona so na primeira semana se não me engano.

Eu queria um tutorial que vi um cara postar na comunidade Front-End do Facebook, de como gravar-se programando

sobre o Rescue Time ,eu uso o forest, li aqui num topico do pessoal sobre as ferramentas que usam, mas querendo ou não depois de umas 2h focado , eu preciso ficar uns 10min, lendo noticias, vendo videos e etc, é pra mim, não sei se é do meu perfil, ou geração meio foda ficar mais que 2,3h focado direto.. 

tambem tenho o costume de ouvir palestras, até cursos ou documentarios no plano de fundo... Parece que eu produzo mais assim.. ou foco até... ou simplesmente trabalho mais animado...

**Comment by danilovaz** [comment]
@rubenmarcus eu estou usando já faz 1 mês e não tive problemas não. o que acontece é que o free você consegue filtrar só por semana e não pegar o mês inteiro que você programou.

**Comment by felipefialho** [comment]
@danilovaz Tinha colocado a key em "default", coloquei em user agora e continua não funcionando.

**Comment by danilovaz** [comment]
@LFeh você gerou a Key aqui? https://wakatime.com/settings/account

No meu eu fiz isso de colocar a Key em user, fechei o sublime e abri de novo. Aí abri o folder do projeto. Não me lembro quanto tempo depois eu fui olhar, mas 1h depois ele já tava contando o tempo que eu ficava codando no sublime.

Estranho, aparentemente tá tudo certo no seu. Era pra já estar pegando, que zica

**Comment by woliveiras** [comment]
Uso o Forest pra não mexer no celular também @rubenmarcus.

Uma coisa que me ajuda a focar (to até preparando um post sobre o assunto) é a técnica de Pomodoro com Pomodoro Clock no celular.

Com Pomodoro a gente tem o tempo de focar nas tarefas, mas também tem os tempos pra fazer coisas não produtivas, o que não faz com que sintamos vontade de ficar indo no Face, Email, Twitter, toda hora. 😄

**Comment by rickbenetti** [comment]
O que não gosto muito do wakatime é que ele pega só o tempo codando e o tempo estudando e analisando o que se está fazendo ou mesmo o tempo de testes em navegadores e isso a maioria esquece.
Por isso estou arrumando a casa pra usar o [Toggl](http://toggl.com/) que fica contando o tempo e eu to pegando a prática de marcar qual a tarefa que estou executando naquele momento.

**Comment by leandroruel** [comment]
link quebrado @rkb81

**Comment by rickbenetti** [comment]
Link corrigido, sorry hábito de não usar mais o http e esqueci do // também rs.

**Comment by felipefialho** [comment]
Usava o Toggl também :)

**Comment by danilovaz** [comment]
@LFeh deu certo aí o seu?

**Comment by lazarofl** [comment]
Precisamos medir a quantidade de tempo que passamos codificando? Já tive essa preocupação mas vejo hoje que uma empresa ou pessoa que se preocupa em cultivar essa cultura perde facilmente a qualidade.

É um excelente exemplo de abordagem homens/hora codificando. Não funciona na pratica, aumentasse a quantidade de bugs, se bonifica por linha de código ao invés de qualidade; cobertura de código; performance e etc.

O pomodoro é bem legal, mas é preciso tomar cuidado para não usar o tempo dele pra justificar a entrega de alguma feature a ser implementada rapidamente, não é linha de produção :)

**Comment by woliveiras** [comment]
Jamais @lazarofl! kkkkk

O Pomodoro só ajuda a focar e ser mais produtivo em menor tempo, mas evito pensar em quantos Pomodoros vou usar pra entregar uma feature, por exemplo.

É legal você comentar, por que eu já fiz esse tipo de coisa cara!

Agora, sobre o que falou de quantificar quanto tempo passamos codando, foi um bom ponto ein. Deu um choque no cérebro aqui.

**Comment by danilovaz** [comment]
@lazarofl Boa observação. É um ponto de vista interessante de  se analisar. Eu acho que nesse caso varia do propósito em que você usa a ferramenta. No meu caso, eu não quero saber quanto tempo passo de fato codando pra mostrar pro chefe, é mais algo pessoal para justamente eu entender a minha realidade e identificar melhorias.

Mas sim, se a pessoa é obrigada a ter essa cultura por questões de cultura empresarial, aí é um problema. Vejo isso como mais aquela caso: o problema não é a ferramenta mas sim como as pessoas a utilizam. Ou seja, uma arma serve tanto para se proteger quanto pra fazer merda, depende do propósito de quem está usando.

Por isso vejo essas ferramentas com mais importância para o próprio desenvolvedor avaliar questões dele mesmo e não para terceiros. Eu imagino que seja benéfico eu saber exatamente quantas horas codei no freela de Fulano. Posso bater isso de frente com minha hora de trabalho e ver se eu eu realmente cobrei o valor que deveria, ou saí perdendo, ou cobrei muito mais do que realmente levei para fazer.

Esse [post aqui](http://lenilsonjr.com/quantificando-seu-tempo-de-programacao-com-o-wakatime/) também é uma boa

**Comment by rickbenetti** [comment]
Sobre o Wakatime eu tenho ele instalado no meu Sublime e em cada projeto eu adicionei inclusive no repositório o arquivo .wakatime-project onde dentro coloco o nome do projeto, assim todos os envolvidos poderão enviar relatórios caso necessário à respeito de X projeto.

Como tenho que prestar esclarecimento de horas realmente utilizadas no desenvolvimento de atividades pra clientes isso é importante, mas como eu mencionei acima o Wakatime não consegue entender o tempo gasto com estudo, planejamento e testes dos projetos, até mesmo tempo dedicado dentro do terminal ele não capta.

Por tanto o uso dele ficou mais restrito para entender o tempo codando em algo efetivamente, mas mais para conhecimento pessoal, e por esta razão estamos mudando pro Toggl, pois eu precisava de algo de baixo custo para contabilizar tarejas por jobs e já que o Jira não entrou na lista de serviços que podemos pagar.

**Comment by lazarofl** [comment]
Entendo @danilovaz.

É justamente no ponto da utilização da ferramenta como forma de analisar alguma métrica pessoal para evoluir ou como forma de monitorar um time que eu estou levantando a questão e colocando em debate argumentos contra pois já vi muitos devs se preocuparem muito com isso, inclusive eu. Acho esse um dos poucos espaços onde podemos trabalhar tão bem questões como essa hoje :v: 

A questão é, com essas ferramentas que analisam a quantidade de código em cada linguagem e o tempo em que nos dedicamos a elas nós conseguimos realmente entender onde podemos evoluir? Eu acredito que seja muito mais um problema do que uma ajuda :tired_face: mesmo com a questão da necessidade de se levantar o tempo em que foi dedicado a um projeto para um job freela ou algo do tipo...

Um grande problema que vejo é que essas ferramentas nos trazem dados quantitativos ao invés de dados qualitativos, ou seja, é muito fácil olhar o gráfico e chegar a uma conclusão errada de que precisamos dedicar mais tempo em uma linguagem X do que Y, ou que precisamos **escrever** mais ou menos em um linguagem X :scream:. Quando entregamos uma feature o cliente sempre analisa questões qualitativas, ou seja, o quanto de valor você entregou para a organização ou projeto durante um período de tempo.

E esse problema da questão de quantidade X qualidade é praticamente uma epidemia dentro das organizações, com sistemas de bonificação atrelados por quantidade de horas, quantidade de features, quantidade de código, quantidade de bugs... Muitas vezes se bonificam pessoas numa sprint caso elas consigam entregar 100% das features sem métricas essenciais de Definition of Done que considerem cobertura de código, performance, qualidade de código medida por ferramentas de code review e etc.

Logo, acho que essas ferramentas apesar de ajudarem a **estimar** o tempo dedicado a codificação não trazem real benefício para o desenvolvimento da pessoa como um todo, podem gerar até o contrário, quanto a questão do uso delas para entregar um report que a empresa exige, bom, aí é assunto pra outro post mas entendo perfeitamente o uso da ferramenta para ajudar nessa necessidade.

**Comment by felipefialho** [comment]
@danilovaz Não funcionou, não sei qual pode ser o problema.

**Comment by felipefialho** [comment]
@danilovaz Era bug no próprio Wakatime. Fizeram um novo release e corrigiram o problema! 

Agora sim vou poder testar, rs

**Comment by cjpatoilo** [comment]
[Wakatime](https://wakatime.com/@cjpatoilo) é bem foda. Há tb o [Codeivate](http://codeivate.com/users/cjpatoilo). Alguém faz uso?

**Comment by wilkerviana** [comment]
Algo muito legal é que ele também integra com o git e te mostra o histórico de commits e quanto tempo você trabalhou em cada um.
Uso há 1 mês, desde que descobri em um post do @willianjusten, e só tem agregado para tangibilizar tempo de codificação e melhorar meus cálculos de prazos de cada projeto!
